### PR TITLE
Let offsetof pseudo-method support Tuples.

### PR DESCRIPTION
**Pull request by hugopl**
For tuples, the index of the element should be used instead of a instance var.

Following up https://forum.crystal-lang.org/t/get-offset-of-a-tuple-item/2844/3

### CODE REVIEW

**Review by straight-shoota — APPROVED** [review]

**Review by bcardiff — COMMENTED** [review]
What happens if the `offsetof(Foo, 1)` with Foo a class is used?

**Comment by hugopl** [comment]
> What happens if the `offsetof(Foo, 1)` with Foo a class is used?

A beautiful crash `cast from Crystal::NumberLiteral to Crystal::InstanceVar failed, at ...crystal/src/compiler/crystal/semantic/main_visitor.cr:2684:21:2684 (TypeCastError)
` :laughing: 

I'll fix that and add a test.

**Comment by hugopl** [comment]
> What happens if the `offsetof(Foo, 1)` with Foo a class is used?

Similar crash with `offset({Int32}, @hey)`, going to add a test case for both cases.

**Review by jhass — APPROVED** [review]

**Comment by hugopl** [comment]
Is there any change to have this into 1.0?

File: src/compiler/crystal/semantic/main_visitor.cr
Comment (asterite) [inline src/compiler/crystal/semantic/main_visitor.cr:2698]: This works, but I think it's better if it's written like this:
```suggestion
      if type && (type.struct? || type.is_a?(TupleInstanceType))
```

File: src/compiler/crystal/syntax/parser.cr
Comment (asterite) [inline src/compiler/crystal/syntax/parser.cr:5665]: Please move this to a constant, otherwise it's allocating an array every time the parser reaches this.

Comment (asterite) [reply src/compiler/crystal/syntax/parser.cr:5665]: Actually, since you do a check on `@token.type` later on this not needed anymore.

File: src/compiler/crystal/syntax/parser.cr
Comment (asterite) [inline src/compiler/crystal/syntax/parser.cr:5668]: This is no longer just an `instance_var`. I suggest renaming this to `offset` or `offset_exp`.

File: src/compiler/crystal/syntax/parser.cr
Comment (asterite) [inline src/compiler/crystal/syntax/parser.cr:5671]: You can do `if @token.type == :INSTANCE_VAR ... elsif @token.type == :NUMBER ... else raise "expecting instance variable or number literal, not ...".

And here you can also check that the number is not a float and give that as a parse error, instead of a semantic error later on.

File: src/docs_pseudo_methods.cr
Comment (asterite) [inline src/docs_pseudo_methods.cr:82]: Could you add an example for a Tuple?

File: src/compiler/crystal/syntax/parser.cr
Comment (hugopl) [reply src/compiler/crystal/syntax/parser.cr:5671]: Yep, I was in doubt where to throw this error, in the parser or later, mainly when I saw that the number token is for integers and floats, TL;DR; I agree, better let it be a parser error.

**Comment by hugopl** [comment]
Hi,

I agree with all suggestions posted by @asterite and was applying them, but when renaming the attribute `@instance_var` from [Offset AST object](https://github.com/crystal-lang/crystal/blob/a4a371bf74598e6cc025d18d8a309a46328718c0/src/compiler/crystal/syntax/ast.cr#L1093) to `offsetof_exp` I noticed that Crystal exposes this in the public API on `Crystal::Macros::OffsetOf` as `instance_var` (that I wanted to rename to `expression`).

I see there are some breaking changes on this pre1.0, but I don't know if some deprecation can be added.... so... what's the best way to proceed?

1. Leave it as is, rename nothing (bad idea IMO).
2. Rename and let it be a breaking change.
3. Rename and add an alias with a deprecation notice.

**Comment by hugopl** [comment]
I addressed the comments and fixed few details that were missing, like some missing parser tests, not setting the token location on NumberLiteral AST node and other stuff that I'm just still learning from the crystal compiler code.

As the question of what to do about `Crystal::Macros::OffsetOf#instance_var` still open I just added another method called `expression` on this in the PR update.

**Comment by asterite** [comment]
@hugopl Just rename it to `offset` (better than `offset_exp`). It will be a breaking change by I highly doubt anyone is using that in macros yet: it was very recently introduced, and I can't think of a use case for using that in macros.

File: src/compiler/crystal/syntax/ast.cr
Comment (asterite) [inline src/compiler/crystal/syntax/ast.cr:1095]: Let's use `offset` instead of `offset_exp` everywhere. I thought that might conflict with something else, but I don't think so.

Sorry for the going and coming with this! 🙏

Comment (hugopl) [reply src/compiler/crystal/syntax/ast.cr]: No prob, I'm learning a lot about the compiler codebase with this simple patch.

**Comment by asterite** [comment]
@hugopl in subsequent PRs you send I would recommend not amending commits: it's much harder to review (now I don't know what changed from the last time I reviewed this) and PRs are always squashed and merged anyway.

**Comment by hugopl** [comment]
Patch updated.
- Now using `offset` instead of `instance_var/offset_exp` in Offset AST. (i.e. it deserve the _breaking change_ label)
- Updated doc of `Crystal::Macros::OffsetOf#offset` (former `Crystal::Macros::OffsetOf#instance_var`).
- Added a parser test to check the exception thrown when offset parameter is not a instance_var neither a tuple.

I think it's good to go now, but tell me if something can be improved.

**Comment by asterite** [comment]
@hugopl Thanks!

I asked this in the forums, but replying here is fine too: what's your specific use case?

**Comment by hugopl** [comment]
Sent a minor update to fix the formatting of offsetof examples in docs.

File: src/compiler/crystal/syntax/parser.cr
Comment (Sija) [inline src/compiler/crystal/syntax/parser.cr:5672]: I'd use a value class here instead.

Comment (hugopl) [reply src/compiler/crystal/syntax/parser.cr:5672]: yes, I missed to pass the `@token` to the `raise` call.

Comment (hugopl) [reply src/compiler/crystal/syntax/parser.cr]: fixed.

File: src/compiler/crystal/syntax/parser.cr
Comment (Sija) [inline src/compiler/crystal/syntax/parser.cr:5669]: I'd use a class name instead.

File: src/compiler/crystal/syntax/parser.cr
Comment (Sija) [inline src/compiler/crystal/syntax/parser.cr:5672]: ditto

Comment (asterite) [reply src/compiler/crystal/syntax/parser.cr:5672]: Just in case, I'm not going to approve this. I don't think this feature is really needed.

**Review by bcardiff — APPROVED** [review]
I think there is a valid use case to abstract the offset in tuples.

Since the layout is also affected by padding (tuples are not marked as packed right now by the compiler) there an abstraction leak to know the right offset.

Also, although tuple memory representation obeys the order of the type arguments, this is something that could change for example to reduce the memory footprint.

File: src/compiler/crystal/syntax/parser.cr
Comment (hugopl) [reply src/compiler/crystal/syntax/parser.cr:5669]: I didn't get it., Do you mean write this as `raise Whatever.new("...")`? I wrote `raise "..."` just to follow the way other places in the parser do when an error happen.

Comment (Sija) [reply src/compiler/crystal/syntax/parser.cr:5669]: No, I mean using a class/type within the error message and not the value, ie.

```suggestion
                 raise "expecting an integer offset, not '#{@token.class}'", @token if @token.number_kind != :i32
```

Comment (asterite) [reply src/compiler/crystal/syntax/parser.cr:5669]: But that's wrong...

Comment (asterite) [reply src/compiler/crystal/syntax/parser.cr:5669]: @token is always a Token

Comment (Sija) [reply src/compiler/crystal/syntax/parser.cr:5669]: But it has a `#kind` property or sth like that, no?

Comment (straight-shoota) [reply src/compiler/crystal/syntax/parser.cr:5669]: `Token#to_s` handles presentation

Comment (Sija) [reply src/compiler/crystal/syntax/parser.cr:5669]: Fair enough, thanks for bearing with me 🙏

**Review by straight-shoota — APPROVED** [review]

**Comment by hugopl** [comment]
> @hugopl in subsequent PRs you send I would recommend not amending commits: it's much harder to review (now I don't know what changed from the last time I reviewed this) and PRs are always squashed and merged anyway.

Sorry, github UI is too poluted and I really just missed this comment, my bad, Next time I wont be amending the changes...  since GitHub can't show the diff between two versions of the same patch.

**Comment by hugopl** [comment]
As I'm not sure if this will be approved someday or in time to 0.36, I think would be useful to create a PR just renaming/deprecating `Crystal::Macros::OffsetOf#instance_var` to `Crystal::Macros::OffsetOf#expression`, so in the future this can be implemented without breaking any API.

What do you guys think? should I create such PR?

**Comment by straight-shoota** [comment]
I think this can be merged for 0.36.0. Thanks for pointing out.

**Comment by asterite** [comment]
0.36.0 will likely be released tomorrow. Please don't merge any more PRs until the next release. Thank you!

**Comment by bcardiff** [comment]
I'm fine merging this on 0.36. But I am no sure sure if @asterite is convinced of the use case. If that is so, this is GTG on my part.

**Comment by asterite** [comment]
I think it's fine to merge this

**Pull request merged**
